ext/nnstreamer/extra/nnstreamer_python3_helper.cc

```diff
@@ -397,72 +398,101 @@ int addToSysPath (const gchar *path)
     info->info[i].type
         = getTensorType ((NPY_TYPES) (((PyArray_Descr *)shape_type)->type_num));
+    Py_SAFEDECREF (shape_type);
 
-    for (gint j = 0; j < PyList_Size (shape_dims); j++) {
+    PyObject *shape_dims = PyObject_CallMethod (tensor_shape, (char *)"getDims", NULL);
+    if (nullptr == shape_dims) {
+      Py_ERRMSG ("parseTensorsInfo() has failed, cannot get the tensor dimension.");
+      return -1;
+    }
+
+    if (!PyList_CheckExact (shape_dims)) {
+      Py_ERRMSG ("parseTensorsInfo() has failed, dimension should be a list.");
+      Py_SAFEDECREF (shape_dims);
+      return -EINVAL;
+    }
+
+    rank = (guint) PyList_Size (shape_dims);
+    if (rank > NNS_TENSOR_RANK_LIMIT) {
+      Py_ERRMSG ("parseTensorsInfo() has failed, max rank of tensor dimension is %d.", NNS_TENSOR_RANK_LIMIT);
+      Py_SAFEDECREF (shape_dims);
+      return -EINVAL;
+    }
+
+    for (j = 0; j < rank; j++) {
       PyErr_Clear();
       PyObject *item = PyList_GetItem (shape_dims, (Py_ssize_t)j);
+      int val = -1;
+
       if (PyErr_Occurred()) {
         PyErr_Print();
+        PyErr_Clear();
         info->info[i].dimension[j] = 0;
         ml_loge
           ("Python nnstreamer plugin has returned dimensions of the %u'th tensor not in an array. Python code should return int-type array for dimensions. Indexes are counted from 0.\n",
           i);
-        ret = -EINVAL;
+        Py_SAFEDECREF (shape_dims);
+        return -EINVAL;
       }
-      PyErr_Clear();
 
       if (PyLong_Check (item)) {
-        info->info[i].dimension[j] = (guint)PyLong_AsUnsignedLong (item);
+        val = (int) PyLong_AsLong (item);
       } else if (PyFloat_Check (item)) {
         /** Regard this as a warning. Don't return -EINVAL with this */
-        info->info[i].dimension[j] = (guint)PyFloat_AsDouble (item);
+        val = (int) PyFloat_AsDouble (item);
         ml_loge
-          ("Python nnstreamer plugin has returned the %d'th dimension value of the %u'th tensor in floating-point type (%f), which is casted as unsigned-int. Python code should return int-type for dimension values. Indexes are counted from 0.\n",
+          ("Python nnstreamer plugin has returned the %u'th dimension value of the %u'th tensor in floating-point type (%f), which is casted as unsigned-int. Python code should return int-type for dimension values. Indexes are counted from 0.\n",
           j, i, PyFloat_AsDouble (item));
       } else {
         info->info[i].dimension[j] = 0;
         ml_loge
-          ("Python nnstreamer plugin has returned the %d'th dimension value of the %u'th tensor neither in integer or floating-pointer. Python code should return int-type for dimension values. Indexes are counted from 0.\n",
+          ("Python nnstreamer plugin has returned the %u'th dimension value of the %u'th tensor neither in integer or floating-pointer. Python code should return int-type for dimension values. Indexes are counted from 0.\n",
```

Comment:
Note: i and j start from 0. 
0'th, 1'th, 2'th, 3'th...

Reply:
updated